### DISCUSSION TITLE
winlibs threads models

### DISCUSSION THREAD

**Discussion by sisyphus** [StrawberryPerl/Perl-Dist-Strawberry#182] [General]
I know that @shawnlaffan has started building the libraries for the upcoming 5.40 release using winlibs gcc-13.2.0-ucrt-mcf release.
I'm all for that - I've been using that same compiler to build and test perl-5.39.x devel releases over the last 10 months and it has been fine.
However, I don't actually perform any threaded programs, so the fact that I've found it to be "fine" perhaps isn't all that meaningful ;-)
All it really means is that the test suite (which includes some threading tests) has been passing all tests.

The thing that's a bit unsettling is that @brechtsanders is not keen to keep on producing the MCF builds because he has found them wanting in a number of his own apps.
See https://github.com/brechtsanders/winlibs_mingw/issues/204#issuecomment-2102178099 .
(As yet, he has released only posix threads builds of gcc-14.1.0 - and the MCF threads builds of gcc-13.2.0 is only at revision 3 whereas the POSIX threads builds are at revision 7.)

For mine, the disappointment is not that he has found problems with MCF threads compilers, but that he therefore wants to stop making them available.
I would argue that the better way to get those problems fixed is to enable (and encourage) more people to use the MCF threads variants.

Anyway ... the main purpose of opening this discussion is to raise the awareness that if SP wants to use MCF threads compilers <b>for the long term</b>, then SP might have to start building its own mingw-w64 toolsets in future.

Currently, MCF threads compilers are also available at https://gcc-mcf.lhmouse.com/. I have used them, and they're fine - but I'm not really sure if they "fit the bill" re SP's requirements.
That is, I don't know about the licensing arrangements, and I found that they included a few libraries that I needed to remove because they got in the way of some things that I was wanting to do.

**Comment by brechtsanders** [comment]
It's not that I don't want to release winlibs builds with MCF threads, but rather that I don't see an actual release version of [MCF Gthread Library](https://github.com/lhmouse/mcfgthread) at this time.

If there is a need for it I could release a MinGW-w64 build of GCC 14.1.0 with MCF threads, but as long as the [MCF Gthread Library](https://github.com/lhmouse/mcfgthread) is not flagged as a release version I wouldn't recommend this for production use.

**Comment by shawnlaffan** [comment]
It's easy enough to switch the builds back to POSIX threads if it is decided that MCF is unsuitable.  However, we'd need to make that decision in the next week or so given perl 5.40 is on schedule.  

Currently I'm keeping with gcc 13.2.  gcc 14 is likely to shake out more CPAN module compilation issues (as gcc 13 did) and these are best handled first on unix-based systems given that's where most of the CPAN development work is done.

**Comment by shawnlaffan** [comment]
> I don't see an actual release version of [MCF Gthread Library](https://github.com/lhmouse/mcfgthread) at this time.

@brechtsanders - Is there a difference between General Availability and a formal release?
https://github.com/lhmouse/mcfgthread/releases/tag/v1.7-ga.1

**Reply by brechtsanders** [reply]
> Is there a difference between General Availability and a formal release?

I can't really answer that, that's a question for [lhmouse](https://github.com/lhmouse/)

**Comment by brechtsanders** [comment]
> However, we'd need to make that decision in the next week or so given perl 5.40 is on schedule.

Just released a GCC 14.1.0 build with MCF threads:
https://github.com/brechtsanders/winlibs_mingw/releases/tag/14.1.0mcf-11.0.1-ucrt-r1

**Comment by sisyphus** [comment]
Thanks for that, @brechtsanders. 
I've grabbed (and used) both the 32-bit and 64-bit packages to build perl.
Both are fine, apart from some issues that I believe are related to gcc-14.1.0 (and not to the mcf threading model).

I'll close this discussion now ... though further comment is welcome..
